File: src/slowql/rules/performance/aggregation.py
"""
Performance Aggregation rules.
"""

from __future__ import annotations

from typing import Any

from sqlglot import exp

from slowql.core.models import Category, Dimension, Issue, Query, Severity
from slowql.rules.base import ASTRule, PatternRule

__all__ = [
    "HavingWithoutGroupByRule",
    "OrderByInSubqueryRule",
    "UnfilteredAggregationRule",
]


class UnfilteredAggregationRule(ASTRule):
    """Detects aggregation without a WHERE clause."""

    id = "PERF-AGG-001"
    name = "Unfiltered Aggregation"
    description = "Detects COUNT(*), SUM(), AVG() without a WHERE clause on SELECT."
    severity = Severity.MEDIUM
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_AGGREGATION

    impact = "Aggregates entire table, expensive on large datasets"
    fix_guidance = "Add WHERE clause to filter rows before aggregation"

    def check_ast(self, query: Query, ast: Any) -> list[Issue]:
        if not query.is_select:
            return []

        has_agg = bool(
            list(ast.find_all(exp.Count))
            or list(ast.find_all(exp.Sum))
            or list(ast.find_all(exp.Avg))
        )

        if has_agg and not ast.find(exp.Where):
            return [
                self.create_issue(
                    query=query,
                    message="Aggregation without WHERE clause scans entire table.",
                    snippet=query.raw[:80],
                )
            ]
        return []


class OrderByInSubqueryRule(PatternRule):
    """Detects ORDER BY inside a subquery or CTE."""

    id = "PERF-AGG-002"
    name = "ORDER BY in Subquery"
    description = "Detects ORDER BY inside a subquery or CTE where it is typically meaningless."
    severity = Severity.LOW
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_AGGREGATION

    pattern = r"\(\s*SELECT\b[^)]+\bORDER\s+BY\b"
    message_template = "ORDER BY in subquery is typically meaningless: {match}"

    impact = "ORDER BY in subquery is meaningless and wastes sort cost"
    fix_guidance = "Remove ORDER BY from subquery unless paired with LIMIT/TOP"


class HavingWithoutGroupByRule(ASTRule):
    """Detects HAVING clause without GROUP BY."""

    id = "PERF-AGG-003"
    name = "HAVING Without GROUP BY"
    description = (
        "HAVING without GROUP BY treats the entire result set as a single "
        "group. This is valid SQL but almost always a mistake — the developer "
        "likely forgot GROUP BY, and the query returns at most one row."
    )
    severity = Severity.MEDIUM
    dimension = Dimension.PERFORMANCE
    category = Category.PERF_AGGREGATION
    dialects = ()

    impact = (
        "Without GROUP BY, HAVING filters the entire table as one group. "
        "The query returns either 0 or 1 rows, which is rarely intended."
    )
    fix_guidance = (
        "Add the missing GROUP BY clause, or use WHERE instead of HAVING "
        "if no aggregation is needed."
    )

    def check_ast(self, query: Query, ast: Any) -> list[Issue]:
        issues = []
        for node in ast.walk():
            if isinstance(node, exp.Select):
                having = node.args.get("having")
                group = node.args.get("group")
                if having and not group:
                    issues.append(self.create_issue(
                        query=query,
                        message="HAVING without GROUP BY — entire result treated as single group.",
                        snippet=query.raw[:80],
                    ))
        return issues

File: slowql-rs/src/rules/performance/aggregation.rs
use crate::models::issue::Category;
use crate::models::{Dimension, Issue, Query, Severity};
use crate::rules::base::Rule;
use once_cell::sync::Lazy;
use regex::Regex;

struct UnfilteredAggregationRule;
impl Rule for UnfilteredAggregationRule {
    fn id(&self) -> &'static str { "PERF-AGG-001" }
    fn name(&self) -> &'static str { "Unfiltered Aggregation" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfAggregation) }
    fn impact(&self) -> &'static str { "Aggregates entire table, expensive on large datasets." }
fn check(&self, query: &Query) -> Vec<Issue> { if !query.is_select() { return Vec::new(); } if query.source_context == "adhoc" || query.source_context.is_empty() { return Vec::new(); } let upper = query.raw_upper(); let has_agg = upper.contains("COUNT(") || upper.contains("SUM(") || upper.contains("AVG("); if has_agg && !upper.contains("WHERE") { let snip = &query.raw[..query.raw.len().min(80)]; return vec![self.build_issue(query, "Aggregation without WHERE clause scans entire table.", snip)]; } Vec::new() } }

struct OrderByInSubqueryRule;
static PAT_ORDER_SUB: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?i)\(\s*SELECT\b[^)]+\bORDER\s+BY\b").unwrap());
impl Rule for OrderByInSubqueryRule {
    fn id(&self) -> &'static str { "PERF-AGG-002" }
    fn name(&self) -> &'static str { "ORDER BY in Subquery" }
    fn severity(&self) -> Severity { Severity::Low }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfAggregation) }
    fn impact(&self) -> &'static str { "ORDER BY in subquery is meaningless and wastes sort cost." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        PAT_ORDER_SUB.find(&query.raw).map(|m| {
            vec![self.build_issue(query, "ORDER BY in subquery is typically meaningless.", m.as_str())]
        }).unwrap_or_default()
    }
}

struct HavingWithoutGroupByRule;
impl Rule for HavingWithoutGroupByRule {
    fn id(&self) -> &'static str { "PERF-AGG-003" }
    fn name(&self) -> &'static str { "HAVING Without GROUP BY" }
    fn severity(&self) -> Severity { Severity::Medium }
    fn dimension(&self) -> Dimension { Dimension::Performance }
    fn category(&self) -> Option<Category> { Some(Category::PerfAggregation) }
    fn impact(&self) -> &'static str { "Without GROUP BY, HAVING filters the entire table as one group." }
    fn check(&self, query: &Query) -> Vec<Issue> {
        let upper = query.raw_upper();
        if upper.contains("HAVING") && !upper.contains("GROUP BY") {
            let snip = &query.raw[..query.raw.len().min(80)];
            return vec![self.build_issue(query, "HAVING without GROUP BY - entire result treated as single group.", snip)];
        }
        Vec::new()
    }
}

pub fn rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(UnfilteredAggregationRule),
        Box::new(OrderByInSubqueryRule),
        Box::new(HavingWithoutGroupByRule),
    ]
}
